crypto/aes/gf128mul.c

```diff
@@ -0,0 +1,471 @@
+/*
+---------------------------------------------------------------------------
+Copyright (c) 1998-2010, Brian Gladman, Worcester, UK. All rights reserved.
+
+The redistribution and use of this software (with or without changes)
+is allowed without the payment of fees or royalties provided that:
+
+  source code distributions include the above copyright notice, this
+  list of conditions and the following disclaimer;
+
+  binary distributions include the above copyright notice, this list
+  of conditions and the following disclaimer in their documentation.
+
+This software is provided 'as is' with no explicit or implied warranties
+in respect of its operation, including, but not limited to, correctness
+and fitness for purpose.
+---------------------------------------------------------------------------
+Issue Date: 20/12/2007
+
+ This file provides fast multiplication in GF(128) as required by several
+ cryptographic authentication modes (see gfmul128.h).
+*/
+
+/*  Speed critical loops can be unrolled to gain speed but consume more memory */
+#if 1
+#  define UNROLL_LOOPS
+#endif
+
+/* The order of these includes matters */
+#include "mode_hdr.h"
+#include "gf128mul.h"
+#include "gf_mul_lo.h"
+
+#if defined( GF_MODE_LL )
+#  define mode   _ll
+#elif defined( GF_MODE_BL )
+#  define mode   _bl
+#elif defined( GF_MODE_LB )
+#  define mode   _lb
+#elif defined( GF_MODE_BB )
+#  define mode   _bb
+#else
+#  error mode is not defined
+#endif
+
+#if defined( GF_MODE_LL) || defined( GF_MODE_LB )
+#  define GF_INDEX(i)  (i)
+#else
+#  define GF_INDEX(i)  (15 - (i))
+#endif
+
+/* A slow field multiplier */
+
+void gf_mul(gf_t a, const gf_t b)
+{   gf_t p[8];
```

Comment:
I think we require explicit initialization even in foreign code.

Reply:
AFAICS we not. A few examples from several different files:
https://github.com/trezor/trezor-firmware/blob/866853e856299d7498699ae765757a470e253536/crypto/aes/aescrypt.c#L141
https://github.com/trezor/trezor-firmware/blob/866853e856299d7498699ae765757a470e253536/crypto/aes/aes_modes.c#L360
https://github.com/trezor/trezor-firmware/blob/866853e856299d7498699ae765757a470e253536/crypto/aes/aeskey.c#L83

Reply:
The above are just examples. There are at least ten times as many cases in the AES code. If the AES code was excluded in 2019, then I don't see any reason to override that decision today.

Reply:
If I remember correctly, the decision was to enforce it in the foreign code as well. Therefore, my comment is compliant with the decision made at that time.

The reason why it wasn't fully accomplished is because I used an automated tool, which unfortunately failed in numerous instances with the AES code.

Reply:
I see that there were some changes to the AES files as part of the 2019 PR. Personally, I am not convinced we should be doing this, not to mention that it is completely out of scope of this PR, but since we already started, let's at least be consistent. https://github.com/trezor/trezor-firmware/pull/3625/commits/10644d4f3d40114554b9d2c6c93a9de0071c43bc